engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,659 @@
+		"new stop height is later than the current one")
+}
+
+// GetStopParameters returns the upcoming stop parameters or nil if no stop is set.
+func (s *StopControl) GetStopParameters() StopParameters {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.stopBoundary.StopParameters
+}
+
+// ShouldExecuteBlock should be called when new block can be executed.
+// The block should not be executed if its height is above or equal to
+// s.stopBoundary.StopBeforeHeight.
+//
+// It returns a boolean indicating if the block should be executed.
+func (s *StopControl) ShouldExecuteBlock(b *flow.Header) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	// don't process anymore blocks if stopped
+	if s.stopped {
+		return false
+	}
+
+	// Skips blocks at or above requested stopHeight
+	// doing so means we have started the stopping process
+	if b.Height < s.stopBoundary.StopBeforeHeight {
+		return true
+	}
+
+	s.log.Info().
+		Msgf("Skipping execution of %s at height %d"+
+			" because stop has been requested %s",
+			b.ID(),
+			b.Height,
+			s.stopBoundary)
+
+	// stopBoundary is now immutable, because it started affecting execution
+	s.stopBoundary.immutable = true
+	return false
+}
+
+// blockFinalized is called when a block is marked as finalized
+//
+// Once finalization reached stopHeight we can be sure no other fork will be valid at
+// this height, if this block's parent has been executed, we are safe to stop.
+// This will happen during normal execution, where blocks are executed
+// before they are finalized. However, it is possible that EN block computation
+// progress can fall behind. In this case, we want to crash only after the execution
+// reached the stopHeight.
+func (s *StopControl) blockFinalized(
+	ctx irrecoverable.SignalerContext,
+	h *flow.Header,
+) {
+	s.Lock()
+	defer s.Unlock()
+
+	// already stopped, nothing to do
+	if s.stopped {
+		return
+	}
+
+	// We already know the ID of the block that should be executed last nothing to do.
+	// Node is stopping.
+	if s.stopBoundary.stopAfterExecuting != flow.ZeroID {
+		return
+	}
+
+	handleErr := func(err error) {
+		s.log.Err(err).
+			Stringer("block_id", h.ID()).
+			Stringer("stop", s.stopBoundary).
+			Msg("Error in stop control BlockFinalized")
+
+		ctx.Throw(err)
+	}
+
+	s.processNewVersionBeacons(ctx, h.Height)
+
+	// we are not at the stop yet, nothing to do
+	if h.Height < s.stopBoundary.StopBeforeHeight {
+		return
+	}
+
+	parentID := h.ParentID
+
+	if h.Height != s.stopBoundary.StopBeforeHeight {
+		// we are past the stop. This can happen if stop was set before
+		// last finalized block
+		s.log.Warn().
+			Uint64("finalization_height", h.Height).
+			Stringer("block_id", h.ID()).
+			Stringer("stop", s.stopBoundary).
+			Msg("Block finalization already beyond stop.")
+
+		// Let's find the ID of the block that should be executed last
+		// which is the parent of the block at the stopHeight
+		header, err := s.headers.ByHeight(s.stopBoundary.StopBeforeHeight - 1)
```

Comment:
This same query will be made for every height after the stop height? 

I guess it's ok, because the finalization is already beyond stop?

This would be common if the node is catching up from far behind, where the finalization has already reached the stop height, but the execution is still far behind, in that case, it will still continue finalized lots of blocks before the execution reaches the stop height, which means we will make the same check again and again here.

Reply:
the check would only be made once when execution is catching up, due to this:

```
s.stopBoundary.stopAfterExecuting = parentID
```

and this

```
if s.stopBoundary.stopAfterExecuting != flow.ZeroID {
return
}
```

This should be ok for now.

I have opened an issue for this: https://github.com/onflow/flow-go/issues/4423. The idea is to consume Version beacon events directly.